File: gen_btagEff.py
#! /usr/bin/env python

import argparse
import ROOT
import array
import sys 

dcsv_med_WP = {"2016":0.6324, "2017":0.4941, "2018":0.4184}

def get_sample_tag_list(year):
	f = open("condor/samples/samples"+year+"_v7.txt",'r')
	lines = f.readlines()
	ret = []
	for line in lines :
		tag = line.split("_13TeV")[0]
		if tag.startswith('#'): continue
		#if not 'WW_' in tag : continue
		ret.append(tag)
	return ret

def process_name_list(process):
	ret = []
	if(process=="ttbar"):
		ret.append("TTJets")
	elif(process=="qcd"):
		ret.append("QCD_")
	elif(process=="wjets"):
		ret.append("WJetsToLNu")
		ret.append("WJetsToQQ")
	elif(process=="other"):
		ret.append("DYJetsToLL_M-50_HT-")
		ret.append("TTTT_")
		ret.append("TTWJetsToLNu_")
		ret.append("TTZToLLNuNu_")
		ret.append("WZ_")
		ret.append("ZZ_")
		ret.append("WW_")
		ret.append("WZZ_")
		ret.append("WWZ_")
		ret.append("ZZZ_")
		ret.append("WWW_")
	elif(process=="signal_m1000"):
		ret.append("mGluino1000")
	elif(process=="signal_m1100"):
		ret.append("mGluino1100")
	elif(process=="signal_m1200"):
		ret.append("mGluino1200")
	elif(process=="signal_m1300"):
		ret.append("mGluino1300")
	elif(process=="signal_m1400"):
		ret.append("mGluino1400")
	elif(process=="signal_m1500"):
		ret.append("mGluino1500")
	elif(process=="signal_m1600"):
		ret.append("mGluino1600")
	elif(process=="signal_m1700"):
		ret.append("mGluino1700")
	elif(process=="signal_m1800"):
		ret.append("mGluino1800")
	elif(process=="signal_m1900"):
		ret.append("mGluino1900")
	elif(process=="signal_m2000"):
		ret.append("mGluino2000")
	elif(process=="signal_m2100"):
		ret.append("mGluino2100")
	else:
		print "Invaild Process Name! \n Usable Process Names : qcd, ttbar, wjets, other, and test"
	return ret

def printProgress(iteration, total, prefix = '', suffix ='', decimals = 1, barLength=100):
	formatStr = "{0:."+str(decimals)+"f}"
	percent = formatStr.format(100*(iteration/float(total)))
	filledLength = int(round(barLength*iteration/float(total)))
	bar = "#"*filledLength+' '*(barLength-filledLength)
	sys.stdout.write('\r%s |%s| %s%s %s' %(prefix, bar,percent, '%', suffix)),
	if iteration == total : 
		sys.stdout.write('\n')
	sys.stdout.flush()


def gen_btagEff(out_file_path, docuts, process, year):
	print process
	ROOT.TH1.SetDefaultSumw2()

	c = ROOT.TChain("tree")
	c.Add("/xrootd_user/yjeong/xrootd/nanoprocessing/"+year+"/processed/*"+process+"*.root")
	print(c.GetEntries())

	dcsv_med = ("medium", dcsv_med_WP[year])

	eta_cuts_comb = [-2.4, 2.4]
	eta_cuts_incl = [0., 2.4]
	pt_cuts_comb = [20., 30., 50., 70., 100., 140., 200., 300., 600., 1000.,]
	pt_cuts_incl = [20., 1000.]
	pt_cuts_comb_central = [20., 1000.]
	flavor_cuts = [-0.5, 3.5, 4.5, 5.5]

	numerator_comb = ROOT.TH3D("btagEfficiency_"+dcsv_med[0]+"_comb", "btagEfficiency_"+dcsv_med[0]+"_comb", len(eta_cuts_comb)-1, array.array('d', eta_cuts_comb), len(pt_cuts_comb)-1, array.array('d', pt_cuts_comb), len(flavor_cuts)-1, array.array('d', flavor_cuts))
	numerator_comb_central = ROOT.TH3D("btagEfficiency_"+dcsv_med[0]+"_comb_central", "btagEfficiency_"+dcsv_med[0]+"_comb_central", len(eta_cuts_comb)-1, array.array('d', eta_cuts_comb), len(pt_cuts_comb_central)-1, array.array('d', pt_cuts_comb_central), len(flavor_cuts)-1, array.array('d', flavor_cuts))
	numerator_incl = ROOT.TH3D("btagEfficiency_"+dcsv_med[0]+"_incl", "btagEfficiency_"+dcsv_med[0]+"_incl", len(eta_cuts_incl)-1, array.array('d', eta_cuts_incl), len(pt_cuts_incl)-1, array.array('d', pt_cuts_incl), len(flavor_cuts)-1, array.array('d', flavor_cuts))

	denominator_comb = numerator_comb.Clone("btagEfficiency_denominator_comb")
	denominator_comb_central = numerator_comb_central.Clone("btagEfficiency_denominator_comb_central")
	denominator_incl = numerator_incl.Clone("btagEfficiency_denominator_incl")

	entry = 0
	num_entries = c.GetEntries()
	for event in c : 
		if entry % (num_entries/10000) == 0 :
			printProgress(entry, num_entries, 'Progress:', 'Complete', 2, 50)
		entry = entry + 1 
		if(not getattr(c,"pass")) : continue
		for ijet in range(len(c.jets_hflavor)):
			flavor = abs(c.jets_hflavor[ijet]) 
			if (flavor == 21) : flavor = 0
			pt     = c.jets_pt[ijet]
			pt     = min(pt,999.99)
			eta    = abs(c.jets_eta[ijet])
			dcsv   = c.jets_dcsvb[ijet]
			if (c.jets_islep[ijet]) : continue
			if (pt<30) : continue
			if (eta>2.4) : continue
			if (not c.jets_id[ijet]) : continue
			denominator_comb.Fill(eta,pt,flavor)
			denominator_comb_central.Fill(eta,pt,flavor)
			denominator_incl.Fill(eta,pt,flavor)
			if (dcsv > dcsv_med[1]):
				numerator_comb.Fill(eta,pt,flavor)
				numerator_comb_central.Fill(eta,pt,flavor)
				numerator_incl.Fill(eta,pt,flavor)
	print '\n'
	print denominator_incl.GetBinContent(1,1,3)
	print denominator_comb_central.GetBinContent(1,1,3)
	tot =0
	for k in range(1,len(flavor_cuts)):
		for i in range(1,len(eta_cuts_comb)):
			for j in range(1,len(pt_cuts_comb)):
				tot = tot + denominator_comb.GetBinContent(i,j,k)
	print tot
	print numerator_comb.GetBinContent(1,4,3)
	print denominator_comb.GetBinContent(1,4,3)
	print numerator_incl.GetBinContent(1,1,3)
	print numerator_comb_central.GetBinContent(1,1,3)
	eff_comb = numerator_comb.Clone("Efficiency_comb")
	eff_comb_central = numerator_comb_central.Clone("Efficiency_comb_central")
	eff_incl = numerator_incl.Clone("Efficiency_incl")
	ROOT.TH3D.Divide(numerator_comb, denominator_comb)
	ROOT.TH3D.Divide(numerator_comb_central, denominator_comb_central)
	ROOT.TH3D.Divide(numerator_incl, denominator_incl)
	out_file = ROOT.TFile(out_file_path.replace(".root","_"+process+"_"+str(year)+".root"), "recreate");
#	eff_comb.Write()
#	eff_comb_central.Write()
#	eff_incl.Write()
	numerator_comb.Write()
	numerator_comb_central.Write()
	numerator_incl.Write()
	doc = ROOT.TNamed("Documentation : this file contains a parametrization in (eta,pt,flavor) for deep CSVv2 b-tagger","");
	doc.Write()
	out_file.Close()

if __name__ == "__main__":
	parser = argparse.ArgumentParser(description = "Computes b-tagging efficiency as a fucntion of pT, eta, and flavor", formatter_class=argparse.ArgumentDefaultsHelpFormatter)
	parser.add_argument("-o", "--out_file", metavar="OUT_FILE", default="btagEfficiency.root", help="Save efficiencies to %(metavar)s")
	parser.add_argument("-c", "--docuts", action="store_true", help="Use all available events, applying only basic filters")
	parser.add_argument("-y", "--year", action="store", default="2016",help="Select the year that you want to make trigger efficiency")
#	parser.add_argument("-p", "--process", action="store", default="*StealthSHH_2t4b_mStop*",help="make process")

	args = parser.parse_args()

	print args.docuts
	list_tags = get_sample_tag_list(args.year)
	if args.docuts : 
		gen_btagEff(args.out_file, args.docuts, "WWW_4F_TuneCUETP8M1", args.year)
	else :
		for process in list_tags:
			gen_btagEff(args.out_file, args.docuts, process, args.year)

File: gen_btagEff.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream> // for stringstream
#include <map>

#include "TROOT.h"
#include "TMath.h"
#include "TFile.h"
#include "TTree.h"
#include "TMath.h"
#include "TSystem.h"
#include "TH2F.h"
#include "TEllipse.h"
#include "TCanvas.h"
#include "TDirectory.h"
#include "TBranch.h"
#include "TString.h"
#include "TLorentzVector.h"
#include "TChain.h"
#include "TObjString.h"
#include "TH3D.h"

// JEC
#include "JetCorrectorParameters.h"
#include "JetCorrectionUncertainty.h"
#include "FactorizedJetCorrector.h"

// btag
#include "BTagCalibrationStandalone.h"

//#include "jetTools.h"
//#include "mcTools.h"
#include "lepTools.h"
#include "inJSON.h"

using namespace std;





void gen_btagEff(TString process, TString year)
{
  TH1::SetDefaultSumw2();

  bool 		pass=true;
  vector<int> 	*jets_hflavor(0);
  vector<float> *jets_pt(0);
  vector<float> *jets_eta(0);
  vector<float> *jets_dcsvb(0);
  vector<bool> 	*jets_islep(0);
  vector<bool> 	*jets_id(0);
  vector<bool>  *jets_hem(0);


  TChain *ch = new TChain("tree");
  //ch->Add("/data3/nanoprocessing/processed_230915_muonsf/"+year+"/"+process+"/*.root");
  ch->Add("/data3/nanoprocessing/processed_241029_signal/"+year+"/"+process+"/*.root");
  cout << "GetEntries: " << ch->GetEntries() << endl;

  ch->SetBranchAddress("jets_hflavor",	&jets_hflavor);
  ch->SetBranchAddress("jets_pt",	&jets_pt);
  ch->SetBranchAddress("jets_eta",	&jets_eta);
  ch->SetBranchAddress("jets_dcsvb",	&jets_dcsvb);
  ch->SetBranchAddress("jets_islep",	&jets_islep);
  ch->SetBranchAddress("jets_id",	&jets_id);
  ch->SetBranchAddress("pass",		&pass);
  ch->SetBranchAddress("jets_hem",      &jets_hem);

  TFile *outputfile = new TFile("Running/btagEfficiency_"+process+"_"+year+".root", "recreate", "btagEfficiency_"+process+"_"+year+".root");      // Name, option, Title
  outputfile->cd();

  map<TString, double> dcsv_med_WP;
  dcsv_med_WP["UL2016_preVFP"] = 0.6001;             // b-jet eff.: 66.7%
  dcsv_med_WP["UL2016"] = 0.5847;                    // b-jet eff.: 66.9%
  dcsv_med_WP["UL2017"] = 0.4506;                    // b-jet eff.: 71.7%
  dcsv_med_WP["UL2018"] = 0.4168;                    // b-jet eff.: 72.9%

  // define cuts
/*
  double eta_cuts_comb[2] 	 = {-2.4, 2.4};
  double eta_cuts_incl[2] 	 = {0., 2.4};
  double pt_cuts_comb[10] 	 = {20., 30., 50., 70., 100., 140., 200., 300., 600., 1000.,};
  double pt_cuts_incl[2] 	 = {20., 1000.};
  double pt_cuts_comb_central[2] = {20., 1000.};
  double flavor_cuts[4] 	 = {-0.5, 3.5, 4.5, 5.5};
*/

  double eta_cuts[13] 	 = {0.0, 0.2, 0.4, 0.6, 0.8, 1.0, 1.2, 1.4, 1.6, 1.8, 2.0, 2.2, 2.4};
  double pt_cuts[10] 	 = {30., 50., 70., 100., 140., 200., 300., 600., 1000.};
  double flavor_cuts[4]	 = {-0.5, 3.5, 4.5, 5.5};

  // define numerator and denominator histograms
  TH3D* numerator;
  TH3D* denominator;

  numerator = new TH3D("btagEfficiency_medium", "btagEfficiency_medium", 12, eta_cuts, 8, pt_cuts, 3, flavor_cuts);
  denominator = static_cast<TH3D*>(numerator->Clone("btagEfficiency_denominator"));

  int flavor=0;
  double pt=0, eta=0, dcsv=0;;
  double tot=0;

  // event loop
  for(int ievt=0; ievt<ch->GetEntries(); ievt++){

    pass=true;
    jets_hflavor->clear();
    jets_pt->clear();
    jets_eta->clear();
    jets_dcsvb->clear();
    jets_islep->clear();
    jets_id->clear();
    jets_hem->clear();
    flavor=0, pt=0, eta=0, dcsv=0, tot=0;

    if(ievt%10000==0) cout << ievt << " / " << ch->GetEntries() << endl;

    ch->GetEntry(ievt);
    if(pass!=1) continue;
    for(int ijet=0; ijet<jets_hflavor->size(); ijet++) {

      flavor=0, pt=0, eta=0, dcsv=0;


      flavor = TMath::Abs(jets_hflavor->at(ijet));
      if(flavor==21) flavor=0;
      if(jets_pt->at(ijet)>999.99) jets_pt->at(ijet)=999.98;
      pt = jets_pt->at(ijet);
      eta = TMath::Abs(jets_eta->at(ijet));
      dcsv = jets_dcsvb->at(ijet);

      if(jets_islep->at(ijet)==1) continue;
      if(pt<30) break;
      if(eta>2.4) continue;
      if(jets_id->at(ijet)!=1) continue;
      if(jets_hem->at(ijet)==1) continue;

      denominator->Fill(eta, pt, flavor);
      if(dcsv > dcsv_med_WP[year]) { 
        numerator->Fill(eta, pt, flavor);
      }
    }
  }

  numerator->Divide(denominator);
  numerator->Write();

  outputfile->Close();
  cout << "... transferring output file" << endl;
  cout << Form("... cp %s /home/bjhong/nanoprocessing/btagEfficiency/ultralegacy/ptetabinned/%s/%s", outputfile->GetName(), year.Data(), outputfile->GetTitle()) << endl;
  gSystem->Exec(Form("cp %s /home/bjhong/nanoprocessing/btagEfficiency/ultralegacy/ptetabinned/%s/%s", outputfile->GetName(), year.Data(), outputfile->GetTitle()));
  cout << Form("... rm %s", outputfile->GetName()) << endl;
  gSystem->Exec(Form("rm %s", outputfile->GetName()));
  cout << "finish generating btag eff file" << endl;
}

int main(int argc, char **argv)
{

  TString year, process;
  year    = argv[1];
  process = argv[2];
  gen_btagEff(process, year);

  return 0;








}
